pkg/aggregator/demultiplexer.go

```diff
@@ -0,0 +1,361 @@
+	cleanSenders()
+}
+
+// AgentDemultiplexer is the demultiplexer implementation for the main Agent
+type AgentDemultiplexer struct {
+	sync.Mutex
+
+	// options are the options with which the demultiplexer has been created
+	options    DemultiplexerOptions
+	aggregator *BufferedAggregator
+	output
+	*senders
+}
+
+// DemultiplexerOptions are the options used to initialize a Demultiplexer.
+// Note that forwarders are started automatically only if `StartForwarders`
+// is set to `true`.
+type DemultiplexerOptions struct {
+	ForwarderOptions           *forwarder.Options
+	NoopEventPlatformForwarder bool
+	NoEventPlatformForwarder   bool
+	NoOrchestratorForwarder    bool
+	FlushInterval              time.Duration
+	StartForwarders            bool // unit tests don't need the forwarders to be instanciated
+}
+
+type forwarders struct {
+	shared        *forwarder.DefaultForwarder
+	orchestrator  *forwarder.DefaultForwarder
+	eventPlatform epforwarder.EventPlatformForwarder
+}
+
+type output struct {
+	forwarders       forwarders
+	sharedSerializer serializer.MetricSerializer
+}
+
+// DefaultDemultiplexerOptions returns the default options to initialize a Demultiplexer.
+func DefaultDemultiplexerOptions(options *forwarder.Options) DemultiplexerOptions {
+	if options == nil {
+		options = forwarder.NewOptions(nil)
+	}
+
+	return DemultiplexerOptions{
+		ForwarderOptions: options,
+		FlushInterval:    DefaultFlushInterval,
+	}
+}
+
+// InitAndStartAgentDemultiplexer creates a new Demultiplexer and runs what's necessary
+// in goroutines. As of today, only the embedded BufferedAggregator needs a separate goroutine.
+// In the future, goroutines will be started for the event platform forwarder and/or orchestrator forwarder.
+func InitAndStartAgentDemultiplexer(options DemultiplexerOptions, hostname string) *AgentDemultiplexer {
+	// prepare the multiple forwarders
+	// -------------------------------
+
+	log.Debugf("Starting forwarders")
+	// orchestrator forwarder
+	var orchestratorForwarder *forwarder.DefaultForwarder
+	if !options.NoOrchestratorForwarder {
+		orchestratorForwarder = buildOrchestratorForwarder()
+	}
+
+	// event platform forwarder
+	var eventPlatformForwarder epforwarder.EventPlatformForwarder
+	if !options.NoEventPlatformForwarder && options.NoopEventPlatformForwarder {
+		eventPlatformForwarder = epforwarder.NewNoopEventPlatformForwarder()
+	} else if !options.NoEventPlatformForwarder {
+		eventPlatformForwarder = epforwarder.NewEventPlatformForwarder()
+	}
+
+	sharedForwarder := forwarder.NewDefaultForwarder(options.ForwarderOptions)
+
+	// prepare the serializer
+	// ----------------------
+
+	sharedSerializer := serializer.NewSerializer(sharedForwarder, orchestratorForwarder)
+
+	// prepare the embedded aggregator
+	// --
+
+	agg := InitAggregatorWithFlushInterval(sharedSerializer, eventPlatformForwarder, hostname, options.FlushInterval)
+
+	// --
+
+	if demultiplexerInstance != nil {
```

Comment:
We have a race-cond here and might miss this warning. We might want to move this line 180 and use a lock, no ?

When would this happens ? Is this a legit use case for the demuxer ? If so, what about 2 methods: Should we have 2 methods: 
- `InitAndStartAgentltiplexer` to create one and return it
- `InitAndStartAgentDemultiplexer to call `InitAndStartAgentltiplexer` and set `demultiplexerInstance`

Reply:
OK I'll adapt this but just some notes:
  * the `demultiplexerInstance` will be removed at some point
  * during the transition time where it is used, there is no reasons to create multiple demuxer, so these extra checks were more about best-effort than anything
  * even in the future, we may never want to instantiate multiple demuxer (but this is the last polint because debatable).